#!/bin/bash

#support human, zebrafish and mouse
species=$1

#1. download protein file from NCBI (CDD_NCBI new folder where protein file will be stored)
python3 NCBICDD_Download.py $species


#2. generate <species>_CDD.Json
# . add manually curated protein domain only for human
# . Will cause error and exit running process if protein_manuallyCurated.json missed from current directory when species is human
# . ISO screening and generate <species>_CDD.Json. Only keep the preferred isoform listed in ISOLIST if provided
    # . ISOLIST file "your prefered isoform accession ID file with one id each row"   
    # . if not ISOLIST file in current directory, all isoform will be included
#add parameter of --cddid to include only domains with cdd id

if [ $species == 'human' ]; then
	python3 CDD_Annotation.py -p CDD_NCBI_$species -s $species --isoform ISOLIST --cddid
else
	python3 CDD_Annotation.py -p CDD_NCBI_$species -s $species --cddid
fi

#3. generate Ensembl_domain_<species>.json
#domain tables including at least one of the following tables is required before running this script.
#CDD, Pfam, SMART and TIGRFAM
#domain table includes ensembl transcript ID, gene name, Domain ID, aa start, aa end
#where to down CDD domain table:
#http://useast.ensembl.org/biomart/martview/0b33cafdaebd19673cc93f70f64b6ccd


DomainFile=("ensembl_CDD_$species.gz" "ensembl_Pfam_$species.gz" "ensembl_smart_$species.gz" "ensembl_tigrfam_$species.gz" "ensembl_CDD_$species" "ensembl_Pfam_$species" "ensembl_smart_$species" "ensembl_tigrfam_$species")
DomainFileCK=0
for i in "${DomainFile[@]}"
do
        if test -f "${i}"; then
                DomainFileCK=1
        fi
done


if [ $DomainFileCK = 1 ]; then
	python3 Ensembl_CDD.py -s $species ensembl_CDD_$species.gz ensembl_Pfam_$species.gz ensembl_smart_$species.gz ensembl_tigrfam_$species.gz
else
	echo 'Error: Ensembl domain tables including at least one of the following tables is required before running this script'
	echo 'CDD, Pfam, SMART and TIGRFAM'
	echo 'http://useast.ensembl.org/biomart/martview/0b33cafdaebd19673cc93f70f64b6ccd'
	echo 'else to run python3 ensembl_download.py -s $species to download the above files (python3 package pybiomart required)' 
	exit
fi


#4. merge domain for refseq and ensembl genes
cat ${species}_CDD.Json Ensembl_domain_$species.json >proteindomain_$species.json

#5. merge mitochondrail domain mt.domain for human only
# "mt.domain" was generated by /home/jwang7/Work/PPDATA/MTadd2PPDB.ipynb
# NM_006060_proteindomain domains from Mullighan lab
if [ $species == 'human' ]; then
	cat mt.domain >>proteindomain_$species.json
	cat NM_006060_proteindomain >>proteindomain_$species.json
fi
#6. sqlite db 
cat proteindomain.sql >proteindomain_$species.sql
echo ".import proteindomain_${species}.json domain" >>proteindomain_$species.sql
echo "CREATE INDEX domain_isoform on domain(isoform collate nocase);" >>proteindomain_$species.sql
sqlite3 proteindomain_$species.db <proteindomain_$species.sql 

rm -f cddid.tbl


