<!DOCTYPE html>
<html>
<head>
	 <meta charset="utf-8">
</head>
<body>

<script>

/**************************************************
Make links for figures for survivorship portal paper
***************************************************/


//////////////// Descriptive statistics plots /////////////////

const dictionary = {
	genome:'hg38',
	dslabel:'SJLife',
	nav: { activeTab: 1 },
	plots: [{
		chartType: 'dictionary'
	}]
}

const barchart = {
	genome:'hg38',
	dslabel:'SJLife',
	nav: { activeTab: 1 },
	plots:[{
		chartType: 'summary',
		term: {id: 'diaggrp'},
		term2: {id: 'agedx'},
		settings: {barchart: {asterisksVisible: false}}
	}]
}

const violin = {
	genome:'hg38',
	dslabel:'SJLife',
	nav: { activeTab: 1 },
	plots:[{
		chartType: 'summary',
		term: {id: 'LV_Ejection_Fraction_3D', q: {mode: 'continuous'}},
		term2: {id: 'agelastvisit'}
	}]
}

const cuminc = {
	genome:'hg38',
	dslabel:'SJLife',
	nav: { activeTab: 1 },
	plots:[{
		chartType: 'cuminc',
		term: {id: 'Cardiomyopathy', q: {breaks: [3]}},
		term2: {
			id: 'anthracyclines_cog_5',
			q: {
				type: "custom-bin",
				mode: "discrete",
				lst: [{
					startunbounded: true,
					stop: 250,
					stopinclusive: false,
					"label": "<250",
				}, {
					start: 250,
					startinclusive: true,
					stopunbounded: true,
					label: "â‰¥250",
				}]
			}
		}
	}]
}


const cuminc2 = {
	genome:'hg38',
	dslabel:'SJLife',
	nav: { activeTab: 1 },
	termfilter: {
		filter:{
			type:'tvslst',
			join:'',
			in:true,
			lst:[
				{type:'tvs',tvs: {
					term:{id: 'genetic_race'},
					values:[
						{key: 'European Ancestry', label: 'European Ancestry'},
						{key: 'African Ancestry', label: 'African Ancestry'}
					]
				}}
			]
		}
	},
	plots:[{
		chartType: 'cuminc',
		term: {id: 'Cardiomyopathy', q: {breaks: [3]}},
		term2: {id: 'genetic_race'},
		term0: {id: 'sex'}
	}]
}


////////////////////// Regression //////////////////////////

const regression = {
	genome:'hg38',
	dslabel:'SJLife',
	nav: { activeTab: 1 },
	plots:[{
		chartType: 'regression',
		regressionType: 'linear',
		outcome: {id: 'BP_SBP1'},
		independent: [
			{id:'sex'},
			{
				id:'agedx',
				q: {
					mode: 'discrete',
					last_bin: {
						start: 15,
						stopunbounded: true
					}
				},
			},
			{
				id:'agelastvisit',
				q: {
					mode: 'discrete',
					last_bin: {
						start: 55,
						stopunbounded: true
					}
				},
				refGrp: '25 to <35'
			}
		]
	}]
}

const snplstTerm = {
	term:{
		type:'snplst',
		snps:[
			//{rsid:'rs201382018',effectAllele:'T'}, // one het case, African sjlife, id:5506
			{rsid:'rs1641548',effectAllele:'T'},
			{rsid:'rs1642793',effectAllele:'C'},
			{rsid:'rs1042522'},
			{rsid:'rs1642782'},
			{rsid:'rs6503048'}
		],
	},
	q:{
		AFcutoff:5,
		alleleType:0,
		geneticModel:0,
		missingGenotype:0,
		restrictAncestry:{
			name:'European ancestry',
			tvs: {
				term: {
					id: 'genetic_race',
					type: 'categorical',
					name: 'Genetically defined race'
				},
				values: [{ key: 'European Ancestry', label: 'European Ancestry' }]
			}
		}
	}
}

const variant_filter = {
	"type":"tvslst",
	"join":"and",
	"in":true,
	"lst":[
		{"type":"tvs","tvs":{"isnot":true,"values":[{"label":"Bad","key":"Bad"}],"term":{"id":"QC","name":"Classification","parent_id":null,"isleaf":true,"type":"categorical","values":{"Good":{"label":"Good","key":"Good"},"Bad":{"label":"Bad","key":"Bad"}}}}},
		{"type":"tvs","tvs":{"ranges":[{"start":0.95,"startinclusive":true,"stopunbounded":true}],"term":{"id":"SJcontrol_CR","name":"SJLIFE control call rate","parent_id":null,"isleaf":true,"type":"float"}}},
		{"type":"tvs","tvs":{"ranges":[{"start":0.95,"startinclusive":true,"stopunbounded":true}],"term":{"id":"CR","name":"Call rate, SJLIFE+CCSS","parent_id":null,"isleaf":true,"type":"float"}}},
		{"type":"tvs","tvs":{"ranges":[{"start":0.95,"startinclusive":true,"stopunbounded":true}],"term":{"id":"CR_sjlife","name":"SJLIFE call rate","parent_id":null,"isleaf":true,"type":"float"}}},
		{"type":"tvs","tvs":{"ranges":[{"start":0.95,"startinclusive":true,"stopunbounded":true}],"term":{"id":"CR_ccss","name":"CCSS call rate","parent_id":null,"isleaf":true,"type":"float"}}},
		//{"type":"tvs","tvs":{"ranges":[{"start":0.95,"startinclusive":true,"stopunbounded":true}],"term":{"id":"gnomAD_CR","name":"gnmoAD call rate","parent_id":null,"isleaf":true,"type":"float"}}},
		{"type":"tvs","tvs":{"ranges":[{"start":0.1,"startinclusive":true,"stopunbounded":true}],"term":{"id":"gnomAD_AF","name":"gnomAD allele frequency","parent_id":null,"isleaf":true,"type":"float","min":0,"max":1,"values":{}}}},
		{"type":"tvs","tvs":{"isnot":true,"values":[{"label":"yes","key":"1"}],"term":{"id":"Polymer_region","name":"Polymer region","parent_id":null,"isleaf":true,"type":"categorical"}}}
	]
}

const snplocusTerm = {
	term: {type: 'snplocus'},
	q: {
		chr:'chr17', start:7661778, stop:7687538,
		AFcutoff:5,
		alleleType:0,
		geneticModel:0,
		restrictAncestry:{
			name:'European ancestry',
			tvs: {
				term: {
					id: 'genetic_race',
					type: 'categorical',
					name: 'Genetically defined race'
				},
				values: [{ key: 'European Ancestry', label: 'European Ancestry' }]
			}
		},
		variant_filter
	}
}

regression_snplst = JSON.parse(JSON.stringify(regression))
regression_snplst.plots[0].independent.push(snplstTerm)
regression_snplocus = JSON.parse(JSON.stringify(regression))
regression_snplocus.plots[0].independent.push(snplocusTerm)


////////////// regression use-case (Yadav et al., 2021) //////////////////

// PMID: 33288658
const yadav2021 = {
	genome:'hg38',
	dslabel:'SJLife',
	nav: { activeTab: 1 },
	termfilter: {
		filter:{
			type:'tvslst',
			join:'',
			in:true,
			lst:[
				{type:'tvs',tvs: {
					term: {
						id: 'anthracyclines_cog_5',
						type: 'float',
						values: {
							"-8888": { "label":"exposed, dose unknown", "uncomputable":true },
							"-9999": { "label":"unknown exposure", "uncomputable":true }
						}
					},
					ranges: [
						{start:0, startinclusive: false, stopunbounded: true}
					]
				}}
			]
		}
	},
	plots:[{
		chartType: 'regression',
		independent: [
			{id:'sex'},
			{
				id:'agedx',
				q: {
					mode: 'discrete',
					last_bin: {
						start: 15,
						stopunbounded: true
					}
				},
			},
			{
				id:'agelastvisit',
				q: {
					mode: 'discrete',
					last_bin: {
						start: 55,
						stopunbounded: true
					}
				},
				refGrp: '25 to <35'
			},
			{
				id:'anthracyclines_cog_5',
				q: {
					mode:'discrete',
					type:'custom-bin',
					lst:[
						{startunbounded:true, stopinclusive:true,stop:0,label:'Not exposed'},
						{stopinclusive:true,start:0,stop:100,label:'>0 to 100'},
						{stopinclusive:true,start:100,stop:250,label:'>100 to 250'},
						{startinclusive:false,stopunbounded:true,start:250,label:'>250'}
					]
				},
				refGrp:'Not exposed'
			},
			{
				id:'hrtavg',
				q: {
					mode:'discrete',
					type:'custom-bin',
					lst:[
						{startunbounded:true, stopinclusive:true,stop:0,label:'Not exposed'},
						{stopinclusive:true,start:0,stop:1000,label:'>0 to 1000'},
						{stopinclusive:true,start:1000,stop:3000,label:'>1000 to 3000'},
						{startinclusive:false,stopunbounded:true,start:3000,label:'>3000'}
					]
				},
				refGrp:'Not exposed'
			}
		]
	}]
}

// snp rsID: rs6689879
const yadav2021_snp1 = {
	term: {type:'snplocus'},
	q: {
		chr:'chr1',
		start:113559000,
		stop:113559200,
		AFcutoff:5,
		alleleType:0,
		geneticModel:0,
		variant_filter
	}
}

const yadav2021_cardio_snp1_afr = getYadav2021({
	plot: {
		regressionType: 'cox',
		outcome: {
			id: 'Cardiomyopathy',
			q: {timeScale:'age', breaks: [3]}
		}
	},
	snp: yadav2021_snp1,
    ancestry: 'afr'
})

const yadav2021_cardio_snp1_eur = getYadav2021({
	plot: {
		regressionType: 'cox',
		outcome: {
			id: 'Cardiomyopathy',
			q: {timeScale:'age', breaks: [3]}
		}
	},
	snp: yadav2021_snp1,
    ancestry: 'eur'
})

const yadav2021_EF_snp1_afr = getYadav2021({
	plot: {
		regressionType: 'linear',
		outcome: {id: 'LV_Ejection_Fraction_3D'}
	},
	snp: yadav2021_snp1,
    ancestry: 'afr'
})

const yadav2021_EF_snp1_eur = getYadav2021({
	plot: {
		regressionType: 'linear',
		outcome: {id: 'LV_Ejection_Fraction_3D'}
	},
	snp: yadav2021_snp1,
    ancestry: 'eur'
})

const yadav2021_log_cardio_snp1_afr = getYadav2021({
	plot: {
		regressionType: 'logistic',
		outcome: {
			id: 'Cardiomyopathy',
			q: {breaks: [3]}
		}
	},
	snp: yadav2021_snp1,
    ancestry: 'afr'
})

const yadav2021_log_cardio_snp1_eur = getYadav2021({
	plot: {
		regressionType: 'logistic',
		outcome: {
			id: 'Cardiomyopathy',
			q: {breaks: [3]}
		}
	},
	snp: yadav2021_snp1,
    ancestry: 'eur'
})


function getYadav2021(config) {
	const obj = JSON.parse(JSON.stringify(yadav2021))
	const plot = obj.plots[0]
	Object.assign(plot, config.plot)
    const snp = JSON.parse(JSON.stringify(config.snp))
    const ancestry = config.ancestry
    if (ancestry == 'afr') {
        snp.q.restrictAncestry = {
            name: 'African ancestry',
            tvs: {
                term: {id: 'genetic_race', type: 'categorical', name: 'Genetically defined race'},
                values: [{key: 'African Ancestry', label: 'African Ancestry'}]
            }
        }
    }
    else if (ancestry == 'eur') {
        snp.q.restrictAncestry = {
            name: 'European ancestry',
            tvs: {
                term: {id: 'genetic_race', type: 'categorical'},
                values: [{key: 'European Ancestry', label: 'European Ancestry'}]
            }
        }
    } else {
        throw 'ancestry unknown'
    }
	plot.independent.push(snp)
	return obj
}


//////////////// Cumulative incidence use-case ////////////////

const neoplasms_rt = {
	genome:'hg38',
	dslabel:'SJLife',
	nav: { activeTab: 1 },
	plots:[{
		chartType: 'cuminc',
		term: {id: 'Secondary Neoplasms', q: {breaks: [3]}},
		term2: {
			id: 'chestrt_yn',
			q: {
				type: "custom-bin",
				mode: "discrete",
				lst: [{
					startunbounded: true,
					stop: 2000,
					stopinclusive: false,
					"label": "<2000",
				}, {
					start: 2000,
					startinclusive: true,
					stopunbounded: true,
					label: "â‰¥2000",
				}]
			}
		}
	}]
}

const neoplasms_rt_sex = {
	genome:'hg38',
	dslabel:'SJLife',
	nav: { activeTab: 1 },
	plots:[{
		chartType: 'cuminc',
		term: {id: 'Secondary Neoplasms', q: {breaks: [3]}},
		term2: {
			id: 'chestrt_yn',
			q: {
				type: "custom-bin",
				mode: "discrete",
				lst: [{
					startunbounded: true,
					stop: 2000,
					stopinclusive: false,
					"label": "<2000",
				}, {
					start: 2000,
					startinclusive: true,
					stopunbounded: true,
					label: "â‰¥2000",
				}]
			}
		},
		term0: {id: 'sex'}
	}]
}

const breast_rt = {
	genome:'hg38',
	dslabel:'SJLife',
	nav: { activeTab: 1 },
	plots:[{
		chartType: 'cuminc',
		term: {id: 'Breast', q: {breaks: [3]}},
		term2: {
			id: 'chestrt_yn',
			q: {
				type: "custom-bin",
				mode: "discrete",
				lst: [{
					startunbounded: true,
					stop: 2000,
					stopinclusive: false,
					"label": "<2000",
				}, {
					start: 2000,
					startinclusive: true,
					stopunbounded: true,
					label: "â‰¥2000",
				}]
			}
		}
	}]
}


/////////////////// Make figures ///////////////////////

const figures = [
	{
		label: 'Figure 1 - SJLIFE and CCSS cohorts',
		panels: [
			{
				label: 'SJLIFE and CCSS cohorts',
				url: window.location.origin + '/?noheader=1&mass=' + '{"genome":"hg38","dslabel":"SJLife"}'
			}
		]
	},
	{
		label: 'Figure 2 - Descriptive statistics',
		panels: [
			{
				label: 'Panel A - dictionary',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(dictionary)
			},
			{
				label: 'Panel B - bar chart - diagnosis group/agedx',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(barchart)
			},
			{
				label: 'Panel C - violin plot - EF/agelastvisit',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(violin)
			},
			{
				label: 'Panel D - cumulative incidence - cardiomyopathy/anthracyclines',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(cuminc)
			},
			{
				label: 'Cumulative incidence - cardiomyopathy/race/sex',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(cuminc2)
			}
		]
	},
	{
		label: 'Figure 3 - Regression analysis',
		panels: [
			{
				label: 'Panel A - regression - only clinical variables',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(regression)
			},
			{
				label: 'Panel B - regression - with variant list',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(regression_snplst)
			},
			{
				label: 'Panel C - regression - with variant locus',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(regression_snplocus)
			}
		]
	},
	{
		label: 'Figure 4 - Data download (in progress)',
		panels: []
	},
	{
		label: 'Figure 5 - Cumulative incidence use case',
		panels: [
			{
				label: 'Panel A - secondary neoplasms/chest RT',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(neoplasms_rt)
			},
			{
				label: 'Panel B - secondary neoplasms/chest RT/sex',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(neoplasms_rt_sex)
			},
			{
				label: 'Panel C - secondary breast neoplasms/chest RT',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(breast_rt)
			}
		]
	},
	{
		label: 'Figure 6 - Genetic association analysis of cardiomyopathy',
		panels: [
			{
				label: 'Panel A - cardiomyopathy snp #1 - african ancestry',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(yadav2021_cardio_snp1_afr)
			},
			{
				label: 'Panel B - cardiomyopathy snp #1 - european ancestry',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(yadav2021_cardio_snp1_eur)
			},
			{
				label: 'Panel C - EF snp #1 - african ancestry',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(yadav2021_EF_snp1_afr)
			},
			{
				label: 'Panel D - EF snp #1 - european ancestry',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(yadav2021_EF_snp1_eur)
			},
			{
				label: 'Logistic - cardiomyopathy snp #1 - african ancestry',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(yadav2021_log_cardio_snp1_afr)
			},
			{
				label: 'Logistic - cardiomyopathy snp #1 - european ancestry',
				url: window.location.origin + '/?noheader=1&mass=' + JSON.stringify(yadav2021_log_cardio_snp1_eur)
			}
		]
	},
	{
		label: 'Figure 7 - Genetic association of ARID5B and Acute lymphoblastic leukemia',
		panels: [
			{
				label: 'Genome browser (need to update settings)',
				url: window.location.origin + '/example.mds2.html'
			}
		]
	}
]


for (const figure of figures) {
	makeFigure(figure)
}

function makeFigure(figure) {
    const figure_div = document.createElement('div')
    figure_div.innerHTML = figure.label
    document.body.appendChild(figure_div)
    figure_div.append(document.createElement('br'))
    for (const panel of figure.panels) {
        const panel_link = document.createElement('a')
        panel_link.innerHTML = panel.label
        panel_link.setAttribute('href', panel.url)
        panel_link.setAttribute('target','_blank')
        figure_div.appendChild(panel_link)
        figure_div.append(document.createElement('br'))
    }
    figure_div.append(document.createElement('br'))
}

</script>
</body>
</html>