<!DOCTYPE html>
<html>
<head>
	 <meta charset="utf-8">
</head>
<body>


<div id='contentDiv' style='margin:10px;display:grid;grid-template-columns:150px auto 80vw;gap:5px;word-break:break-word'>

-- How it works --

By running makeURL(), rows of URLs are added to this DIV.
This message is also erased.
URL host is window.location.origin, allowing ppr.stjude.org/url.html to work to test things there.

</div>


<!-- search "main" to jump to URL sections -->


<script>

// list of section names, use these to quickly find url contents for a section
const section_BAM = 'BAM',
	section_GDCbam = 'GDC BAM UI',
	section_GDCmds3 = 'GDC lollipop (mds3)',
	section_mds3 = 'mds3 tk',
	section_splicing = 'RNA splicing',
	section_hic = 'HiC, interaction',
	section_bw = 'bigwig',
	section_bedj = 'bedj',
	section_miscTk = 'Misc track types',
	section_study = 'legacy study',
	section_mds = 'mds',
	section_legacyds = 'legacy ds', // to be replaced by mds3
	section_scatterplot = 'scatterplot',
	section_singleCell = 'single cell',
	section_termdbTree = 'termdb tree',
	section_mass = 'MASS UI',
	section_regression = 'regression'


///////////// helpers ////////////////////
const snplstTerm = {
	term:{
		type:'snplst',
		snps:[
			{rsid:"rs1333261826",effectAllele:"T"}, // AF=0, not present in sjlife/ccss
			//{rsid:'rs201382018',effectAllele:'T'}, // one het case, African sjlife, id:5506
			{rsid:'rs1010330999',effectAllele:'G'}, // two het cases, European, sjlife and ccss
			{rsid:'rs1641548',effectAllele:'T'},
			{rsid:'rs858528',effectAllele:'T'},
			{rsid:'rs1642793',effectAllele:'C'},
			{rsid:'rs1042522'},
			{rsid:'rs1642782'},
			{rsid:'rs6503048'},
			// following snps are not in bcf file
			{rsid:'rs1472752888',effectAllele:'T'}
		],
	},
	q:{
		AFcutoff:5,
		alleleType:0,
		geneticModel:0,
		missingGenotype:0,
		restrictAncestry:{
			name:'European ancestry',
			tvs: {
				term: {
					id: 'genetic_race',
					type: 'categorical',
					name: 'Genetically defined race'
				},
				values: [{ key: 'European Ancestry', label: 'European Ancestry' }]
			}
		/*
			name:'African ancestry',
			tvs: {
				term: {
					id: 'genetic_race',
					type: 'categorical',
					name: 'Genetically defined race'
				},
				values: [{ key: 'African Ancestry', label: 'African Ancestry' }]
			}
			*/
		},
	}
}

// 7665606.C.G,T is multi-allelic 
// in european, G is 0%, T and C are common alleles
const snplocusQ = {
	chr:'chr17', start:7664218, stop:7671018,
	AFcutoff:5,
	alleleType:0,
	geneticModel:0,
	restrictAncestry:{
		name:'European ancestry',
		tvs: {
			term: {
				id: 'genetic_race',
				type: 'categorical',
				name: 'Genetically defined race'
			},
			//values: [{ key: 'African Ancestry', label: 'African Ancestry' }]
			values: [{ key: 'European Ancestry', label: 'European Ancestry' }]
		}
	},
	variant_filter: {
		"type":"tvslst",
		"join":"and",
		"in":true,
		"lst":[
			{"type":"tvs","tvs":{"isnot":true,"values":[{"label":"Bad","key":"Bad"}],"term":{"id":"QC","name":"Classification","parent_id":null,"isleaf":true,"type":"categorical","values":{"Good":{"label":"Good","key":"Good"},"Bad":{"label":"Bad","key":"Bad"}}}}},
			{"type":"tvs","tvs":{"ranges":[{"start":0.95,"startinclusive":true,"stopunbounded":true}],"term":{"id":"SJcontrol_CR","name":"SJLIFE control call rate","parent_id":null,"isleaf":true,"type":"float"}}},
			{"type":"tvs","tvs":{"ranges":[{"start":0.95,"startinclusive":true,"stopunbounded":true}],"term":{"id":"CR","name":"Call rate, SJLIFE+CCSS","parent_id":null,"isleaf":true,"type":"float"}}},
			{"type":"tvs","tvs":{"ranges":[{"start":0.95,"startinclusive":true,"stopunbounded":true}],"term":{"id":"CR_sjlife","name":"SJLIFE call rate","parent_id":null,"isleaf":true,"type":"float"}}},
			{"type":"tvs","tvs":{"ranges":[{"start":0.95,"startinclusive":true,"stopunbounded":true}],"term":{"id":"CR_ccss","name":"CCSS call rate","parent_id":null,"isleaf":true,"type":"float"}}},
			//{"type":"tvs","tvs":{"ranges":[{"start":0.95,"startinclusive":true,"stopunbounded":true}],"term":{"id":"gnomAD_CR","name":"gnmoAD call rate","parent_id":null,"isleaf":true,"type":"float"}}},
			//{"type":"tvs","tvs":{"ranges":[{"start":0.1,"startinclusive":true,"stopunbounded":true}],"term":{"id":"gnomAD_AF","name":"gnomAD allele frequency","parent_id":null,"isleaf":true,"type":"float","min":0,"max":1,"values":{}}}},
			{"type":"tvs","tvs":{"isnot":true,"values":[{"label":"yes","key":"1"}],"term":{"id":"Polymer_region","name":"Polymer region","parent_id":null,"isleaf":true,"type":"categorical"}}}
		]
	}
}

function __barchart(t1,t2) {
	const chart = { chartType:'barchart', term:t1 }
	if(t2) chart.term2 = t2
	return '?noheader=1&mass='+JSON.stringify( { dslabel:'SJLife', genome:'hg38', plots: [ chart ], nav: { activeTab: 1 } })
}
function __summary(t1,t2) {
	const chart = { chartType:'summary', term:t1 }
	if(t2) chart.term2 = t2
	return '?noheader=1&mass='+JSON.stringify( { dslabel:'SJLife', genome:'hg38', plots: [ chart ], nav: { activeTab: 1 } })
}
function __cuminc({t1,t2,t0,grade,hidden}) {
	const chart = {
		chartType:'cuminc',
		settings:{ cuminc:{ } },
		term: t1
	}
	if(!chart.term.q) chart.term.q = {}
	if(t2) chart.term2 = t2
	if(t0) chart.term0 = t0
	if(grade) chart.term.q.breaks = [grade]
	if(hidden) chart.settings.cuminc.hidden=hidden
	return '?noheader=1&mass='+JSON.stringify( {
		dslabel:'SJLife',
		genome:'hg38',
		plots: [ chart ],
		nav: { activeTab: 1 }
	})
}






///////////// main ////////////////////

/*
call the makeURL() to make URLs
argument is an array; each element defines a set of example URLs showing in a row

{
	name: string
	links: [ 'bam', [ 'bam1', '?bamfile=xx' ], ... ]
	   each element is a text or array of length 2
	   text: create <span> to display text, no url
	   array: [0]: URL label
			  [1]: URL path name
}

*/
makeURL([


{ name: section_BAM,
	card: 'bam',
	links: [
		[ 'TP53 18bp del', '?genome=hg19&block=1&position=chr17:7578191-7578591&bamfile=Test,proteinpaint_demo/hg19/bam/TP53_del.bam' ],
		[ 'k-mer', '?genome=hg19&block=1&position=chr17:7578191-7578591&bamfile=Test,proteinpaint_demo/hg19/bam/TP53_del.bam&variant=chr17.7578383.AGCAGCGCTCATGGTGGGG.A' ],

		['GATA3 complex mutation','?genome=hg19&block=1&bamfile=Test,proteinpaint_demo/hg19/bam/gata3.complex.bam&position=chr10:8100668-8100707'],
		['k-mer','?genome=hg19&block=1&bamfile=Test,proteinpaint_demo/hg19/bam/gata3.complex.bam&position=chr10:8100668-8100707&variant=chr10.8100686.CAC.CCCTGCCTGTTGTGAGCTGCTCTACGTGCCCTACGTGCT'],

		['chr7 3mer expansion','?genome=hg19&block=1&bamfile=Test,proteinpaint_demo/hg19/bam/chr7.3mer.insert.bam&position=chr7:35473-35473&hlregion=chr7:35472-35472'],
		['k-mer','?genome=hg19&block=1&bamfile=Test,proteinpaint_demo/hg19/bam/chr7.3mer.insert.bam&position=chr7:35473-35473&hlregion=chr7:35472-35472&variant=chr7.35473.T.TGCTGCTGCTGCC'],

		['EXOC4 2mer deletion','?genome=hg38&block=1&bamfile=Test,proteinpaint_demo/hg19/bam/intron.2mer.del.bam&position=chr7:133787365-133787395&hlregion=chr7:133787361-133787361'],
		['k-mer','?genome=hg38&block=1&bamfile=Test,proteinpaint_demo/hg19/bam/intron.2mer.del.bam&position=chr7:133787365-133787395&hlregion=chr7:133787361-133787361&variant=chr7.133787370.TGTGTGTGTGTGTGTGTGTGAGAT.TGAT'],

		['SETD2 splice site insertion','?genome=hg19&block=1&bamfile=Test,proteinpaint_demo/hg19/bam/splice_junction_variant.bam&position=chr3:47155334-47155420'],
		['kmer','?genome=hg19&block=1&bamfile=Test,proteinpaint_demo/hg19/bam/splice_junction_variant.bam&position=chr3:47155334-47155420&variant=chr3.47155366.G.GGGGCT'],

		[ 'KIT 3bp del', '?genome=hg19&block=1&position=chr4:55589607-55590007&bamfile=Test,proteinpaint_demo/hg19/bam/kit.exon8.del.bam' ],
		[ 'k-mer','?genome=hg19&block=1&position=chr4:55589607-55590007&bamfile=Test,proteinpaint_demo/hg19/bam/kit.exon8.del.bam&variant=chr4.55589771.ACGA.A'],

		[ 'RNA 8bp insert','?genome=hg19&block=1&bamfile=test,proteinpaint_demo/hg19/bam/rna.8bp.insertion.bam&position=chr11:119155611-119155851'],
		['k-mer','?genome=hg19&block=1&bamfile=test,proteinpaint_demo/hg19/bam/rna.8bp.insertion.bam&position=chr11:119155611-119155851&variant=chr11.119155746.T.TTGACCTGG'],

		['KIT complex mutation','?genome=hg19&block=1&bamfile=test,proteinpaint_demo/hg19/bam/kit.complex.bam&position=chr4:55589768-55589770'],
		['k-mer','?genome=hg19&block=1&bamfile=test,proteinpaint_demo/hg19/bam/kit.complex.bam&position=chr4:55589768-55589770&variant=chr4.55589768.CTTACGA.AGGG'],

		['BCORL1 complex mutation', '?genome=hg19&block=1&bamfile=test,proteinpaint_demo/hg19/bam/bcorl1.complex.bam&position=chrX:129150009-129150071'],
		['k-mer','?genome=hg19&block=1&bamfile=test,proteinpaint_demo/hg19/bam/bcorl1.complex.bam&position=chrX:129150009-129150071&variant=chrX.129150027.GGAAAGCGTAGGGGTCTTTGCTTGCAAGAACAA.GT'],

		['exon deletion','?genome=hg19&block=1&bamfile=test,proteinpaint_demo/hg19/bam/crebbp.bam&position=chr16:3800245-3803429'],
		['ETV6 SV', '?genome=hg19&block=1&bamfile=test,proteinpaint_demo/hg19/bam/etv6runx1.bam&position=chr12:12023245-12023802'],
		['CBFB inversion','?genome=hg19&block=1&bamfile=test,proteinpaint_demo/hg19/bam/CBFB-MYH11.bam&position=chr16:67115815-67117415'],
		['long read, no qual','?genome=hg38&block=1&bamfile=test,proteinpaint_demo/hg38/bam/control.BCL11A-218.bam&position=chr2:60450049-60563872'],
		['FLAG 0/16','?genome=hg19&block=1&position=chr12:25380685-25381606&bamfile=flag,proteinpaint_demo/hg19/bam/flagproblem.bam'],
		'ERR:',
		['missing','?genome=hg19&block=1&bamfile=flag,path/to/missing.bam'],
		['text','?genome=hg19&block=1&bamfile=flag,proteinpaint_demo/hg19/junction/rnapeg'],
		['gz','?genome=hg19&block=1&bamfile=flag,proteinpaint_demo/hg19/junction/file.gz'],
		['no index','?genome=hg19&block=1&bamfile=test,proteinpaint_demo/hg19/bam/noIndex.bam'],
		['wrong index','?genome=hg19&block=1&bamfile=test,proteinpaint_demo/hg19/bam/wrongIndex.bam'],
	]
},




{ name:section_GDCbam,
	card: 'gdcbam',
	links:[
		['GBM EGFR by case id','?gdcbamslice=1&gdc_id=TCGA-06-0211&gdc_pos=chr7:55153818-55156225'],
		['KRAS G12V by case uuid','?gdcbamslice=1&gdc_id=9a2a226e-9605-4214-9320-469305e664e6&gdc_var=chr12.25245350.C.A'],
		['IDH1 R132H by file id','?gdcbamslice=1&gdc_id=00493087-9d9d-40ca-86d5-936f1b951c93_wxs_gdc_realn.bam&gdc_var=chr2:g.208248388C>T'], // chr2.208248388.C.T

		['AKT1 E17K file uuid','?gdcbamslice=1&gdc_id=35918f42-c424-48ef-8c95-2d87b48fdf41&gdc_var=chr14:g.104780214C>T'], // chr14.104780214.C.T

		['bam link','?genome=hg38&block=1&position=chr14:104780136-104780318&variant=chr14.104780214.C.T&bamfile=xx,yy&gdc=35918f42-c424-48ef-8c95-2d87b48fdf41,<yourTokenHere>'],
		['KRAS G12D','?gdcbamslice=1&gdc_id=17284ec9-c3cb-4c80-96a7-312e35e7c1d9&gdc_var=chr12.25245350.C.T'],
		['NPM1 exon12 insertion','?gdcbamslice=1&gdc_id=09c3ee18-6f93-448f-85b6-344be06f8851&gdc_var=chr5:g.171410539_171410540insTCTG'], // chr5.171410539.C.CTCTG

		['MYC 3bp del','?gdcbamslice=1&gdc_id=c3c5b27c-106b-457b-9bdf-faea907301d1_wxs_gdc_realn.bam&gdc_var=chr8:g.127740427delTGT'], // chr8.127740426.ATGT.A

		['GATA3 X351 splice del', '?gdcbamslice=1&gdc_id=df82226e-2242-418b-9f5f-0a5e531826a4_wxs_gdc_realn.bam&gdc_var=chr10:g.8073734delTTTAGA'],

		['ACVR2A K437fs','?gdcbamslice=1&gdc_id=fe42ed1d-e913-4214-a3a4-7baf5288dbb7_wxs_gdc_realn.bam&gdc_var=chr2:g.147926117delA'], // chr2.147926116.TA.T

		['PTPN4 X605 mnv','?gdcbamslice=1&gdc_id=8577ef8a-f071-49cc-8158-05dfc52f3c70&gdc_var=chr2:g.119955155_119955159delinsTTTTT'],

		['Skipped by workflow','?gdcbamslice=1&gdc_id=e2d8ca95-6a3d-48e3-abfc-ce5ee294e954'],
		['6 ssm','?gdcbamslice=1&gdc_id=TCGA-4V-A9QQ'],
		['No ssm','?gdcbamslice=1&gdc_id=DLBCL10969'],
		['disableSSM','?gdcbamslice=1&disableSSM=1']
		// TODO hypermutator
	]
},




{name:section_GDCmds3,
	card:'lollipop',
	links:[
		['AKT1','?genome=hg38&gene=ENST00000407796&mds3=GDC'],
		['NM_005163','?genome=hg38&gene=NM_005163&mds3=GDC'],
		['TP53','?genome=hg38&gene=ENST00000269305&mds3=GDC'],
		['KRAS','?genome=hg38&gene=ENST00000256078&mds3=GDC'],
		['IDH1','?genome=hg38&gene=ENST00000345146&mds3=GDC'],
		['KRAS G12V','example.gdc.byssm.html'],
		['SRP9 small gene','?genome=hg38&gene=ENST00000650651&mds3=GDC'],
		['AKT1 lncRNA','?genome=hg38&gene=ENST00000557552&mds3=GDC'], // no ssm but has cases??
		['IGKC','?genome=hg38&gene=IGKC&mds3=GDC'],
	]
},

{name:section_mds3,
	card:'lollipop',
	links:[
		['ASH', '?genome=hg38&gene=bcr&mds3=ASH'],
		['invalid ds', '?genome=hg38&gene=bcr&mds3=ash'],
		['clinvar', '?genome=hg38&gene=bcr&mds3=clinvar']
		// todo vcf file
	]
},

{name:section_splicing,
	card: 'junction',
	links:[
		['single','?block=on&genome=hg19&junctionfile=RNA%20splice%20junction,proteinpaint_demo/hg19/junction/file.gz'],
		['rnapeg','?genome=hg19&block=1&junctionrnapeg=RNApeg,proteinpaint_demo/hg19/junction/rnapeg'],
		['official mds','?genome=hg19&block=1&mds=Pediatric2,junction'],
		['custom mds json','?genome=hg19&block=1&mdsjunctionfile=Demo,proteinpaint_demo/hg19/junction/json.gz'],
		['custom matrix','?genome=hg19&block=1&junctionmatrix=test,proteinpaint_demo/hg19/junction/matrix.gz']
	]
},
{name:section_hic,
	card: 'hic',
	links:[
		['hic genome','?genome=hg19&hicfile=proteinpaint_demo/hg19/hic/hic_demo.hic&enzyme=MboI'],
		['hic tk file','?block=1&genome=hg19&position=chr7:13749862-20841903&hictkfile=Test_HiC,MboI,proteinpaint_demo/hg19/hic/hic_demo.hic'],
		['hic tk url','?block=1&genome=hg19&position=chr7:13749862-20841903&hictkurl=Test_HiC,MboI,https://proteinpaint.stjude.org/ppdemo/hg19/hic/hic_demo.hic'],
		['bed arc','?genome=hg19&block=1&position=chr11:7839893-9165172&arcfile=Arc%20Track,proteinpaint_demo/hg19/arc/mango.gz'],
	]
},

{name:section_bw,
	card:'bigwig',
	links:[
		['local','?block=on&genome=hg19&bigwigfile=local,proteinpaint_demo/hg19/bigwig/file.bw'],
		['local blank region','?block=on&genome=hg19&bigwigfile=local,proteinpaint_demo/hg19/bigwig/file.bw&position=chr17:7625542-7640838'],
		['prp1 tiny file','?block=on&genome=hg19&bigwigurl=prp1,https://proteinpaint.stjude.org/ppdemo/hg19/bigwig/file.bw'],
		['prp1 big file','?block=on&genome=hg19&bigwigurl=prp1,https://proteinpaint.stjude.org/ppSupport/hg19.100way.phyloP100way.bw'],
		['ucsc','?block=on&genome=hg19&bigwigurl=ucsc,http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/wgEncodeCrgMapabilityAlign100mer.bigWig'],
		['wrongLocal','?block=on&genome=hg19&bigwigfile=local,proteinpaint_demo/hg19/missing.bw'],
		['wrongUrl','?block=on&genome=hg19&bigwigurl=ucsc,http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/missing.bigWig'],
	]
},

{name:section_bedj,
	card: 'bedj',
	links:[
		['bedj','?block=on&genome=hg19&bedjfile=JSON-BED,proteinpaint_demo/hg19/bedj/gencode.gz'],
		['bed3','?genome=hg19&block=1&bedjfile=BED3_Example,proteinpaint_demo/hg19/bedj/bed3.gz'],
		['bigbed gencode','?genome=hg38&block=1&bedjfile=bigbed,proteinpaint_demo/hg38/bedj/gencodeV41.bb'],
		['bigbed dbsnp','?block=on&genome=hg19&position=chr17:7579489-7579868&bedjfile=dbsnp,anno/dbsnp.hg19.bb'],
		['pecan bedj','?genome=hg19&block=1&bedjurl=bed,https://pecan.stjude.cloud/static/hg19/refGene.hg19.gz'],
		['prp1 bedj','?genome=hg19&block=1&bedjurl=bed,https://proteinpaint.stjude.org/ppdemo/hg19/bedj/gencode.gz'],
		// TODO add examples using bb url
		'ERR:',
		['missing','?block=on&genome=hg19&bedjfile=JSON-BED,path/to/missing.gz'],
		['text file','?genome=hg19&block=1&bedjfile=flag,proteinpaint_demo/hg19/junction/rnapeg'],
		['bam','?genome=hg19&block=1&position=chr12:25380685-25381606&bedjfile=flag,proteinpaint_demo/hg19/bam/flagproblem.bam'],
		['no index','?genome=hg19&block=1&bedjfile=no%20index,proteinpaint_demo/hg19/bedj/noIndex.gz'],
		['wrong index','?genome=hg19&block=1&bedjfile=wrong%20index,proteinpaint_demo/hg19/bedj/wrongIndex.gz'],
	]
},

{name:section_miscTk,
	links:[
		['aicheck','?block=on&genome=hg19&position=chr9:36816442-37058548&aicheckfile=Allelic%20imbalance,proteinpaint_demo/hg19/aicheck/file.gz'],
		['bampile','?block=on&genome=hg19&position=chr12:25362480-25363264&bampilefile=Test,proteinpaint_demo/hg19/bampile/tk.gz'],
	]
},

{name:section_study,
	card: 'study',
	links:[
		['ERG','?study=proteinpaint_demo/hg19/heatmap/heatmapDemo.json'],
	]
},

{name:section_legacyds,
	links:[
		['pediatric','?genome=hg19&gene=bcr&dataset=pediatric'],
		['cosmic','?genome=hg19&gene=bcr&dataset=COSMIC'],
	]
},

{name:section_mds,
	card:'genomepaint',
	links:[
		['pediatric','?genome=hg19&block=1&mds=Pediatric2,svcnv'],
		['tcga-dlbc','?genome=hg38&block=1&position=chr22:22569655-23013766&svcnvfpkmfile=TCGA_DLBC,svcnv,proteinpaint_demo/hg38/GP/TCGA_DLBC.CNV.gz,vcf,proteinpaint_demo/hg38/GP/TCGA_DLBC.vcf.gz,fpkm,proteinpaint_demo/hg38/GP/TCGA_DLBC.fpkm.gz'],
		['mdsjson','?genome=hg38&block=1&position=chr22:22569655-23013766&mdsjson=proteinpaint_demo/hg38/GP/mdsjson/mdsjsondemo.json']
	]
},

{name:section_scatterplot,
	card: 'scatterplot',
	links:[
		['sjcloud', '?scatterplot=1&genome=hg38&tsnejson=proteinpaint_demo/hg38/mdstsne/file.json']
	]
},
{name:section_singleCell,
	card: 'singlecell',
	links:[
		['2D','?genome=hg19&singlecell=proteinpaint_demo/hg19/singlecell/singlecell.json']
	]
},

{name:section_termdbTree,
	card:'databrowser',
	links:[
		['TermdbTest', '?termdb={"vocab":{"dslabel":"TermdbTest","genome":"hg38"}}'],
		['SJLife', '?termdb={"vocab":{"dslabel":"SJLife","genome":"hg38"}}'],
		['ASH', '?termdb={"vocab":{"dslabel":"ASH","genome":"hg38"}}'],
		['GDC', '?termdb={"vocab":{"dslabel":"GDC","genome":"hg38"}}'],
		['MSigDB', '?termdb={"vocab":{"dslabel":"msigdb","genome":"hg38"}}'],
	]
},

{name:section_mass,
	links:[
		['TermdbTest', '?noheader=1&mass={"genome":"hg38","dslabel":"TermdbTest"}'],
		['SJLife', '?noheader=1&mass={"genome":"hg38","dslabel":"SJLife"}'],
		['CCSS', '?noheader=1&mass={"genome":"hg38","dslabel":"SJLife","activeCohort":1}'],
		['PNET', '?noheader=1&mass={"genome":"hg19","dslabel":"PNET"}'],
		['ASH', '?noheader=1&mass={"genome":"hg38","dslabel":"ASH"}'],
		['ALL-pharmacotyping', '?noheader=1&mass={"genome":"hg38","dslabel":"ALL-pharmacotyping"}'],

		'barchart',

		['diaggrp', __barchart({ id: 'diaggrp' })],
		['diaggrp/race', __barchart( { id: 'diaggrp' }, { id: 'racegrp' })],
		['diaggrp/hrtavg', __barchart( { id: 'diaggrp' }, { id: 'hrtavg' })],
		['diaggrp/Neurology', __barchart( { id: 'diaggrp' },  { id: 'Neurology System' })],
		['diaggrp/Neurology break2', __barchart( { id: 'diaggrp' },
			{ id: 'Neurology System', q:{breaks:[2], groupNames:['No condition/not tested','Has condition']} })
		],

		['hrtavg', __barchart( { id: 'hrtavg' })],
		['hrtavg/diaggrp', __barchart( { id: 'hrtavg' }, { id: 'diaggrp' })],
		['hrtavg/agedx', __barchart( { id: 'hrtavg' }, { id: 'agedx' })],
		['hrtavg/Neurology', __barchart( { id: 'hrtavg' }, { id: 'Neurology System' })],
		['hrtavg/Neurology break2', __barchart( { id: 'hrtavg' },
			{ id: 'Neurology System', q:{breaks:[2], groupNames:['No condition/not tested','Has condition']} })
		],

		['Neurology', __barchart( { id: 'Neurology System' })],
		['Neurology/diaggrp', __barchart( { id: 'Neurology System' }, { id: 'diaggrp' })],
		['Neurology/hrtavg', __barchart( { id: 'Neurology System' }, { id: 'hrtavg' })],
		['Neurology/Cardio', __barchart( { id: 'Neurology System' }, { id: 'Cardiovascular System' })],

		'violin',
		['hrtavg', __barchart( { id: 'hrtavg', q:{mode:'continuous'} })],
		['hrtavg/race', __barchart( { id: 'hrtavg', q:{mode:'continuous'} }, {id:'racegrp'})],
		

		'summary',
		['hrtavg', __summary( { id: 'hrtavg', q:{mode:'continuous'} })],
		['hrtavg/race', __summary( { id: 'hrtavg', q:{mode:'continuous'} }, {id:'racegrp'})],


		'cuminc',

		['Cardio-1', __cuminc({t1:{id: 'Cardiovascular System'},grade:1})],
		['Cardio-3', __cuminc({t1:{id: 'Cardiovascular System'},grade:3})],
		['Cardio/sex', __cuminc({t1:{id: 'Cardiovascular System'},t2:{id:'sex'}})],
		['Cardio/ hrtavg', __cuminc({
			t1:{id: 'Cardiovascular System'},
			t2:{id:'hrtavg', q: {
				type: "custom-bin",
				mode: "discrete",
				lst: [{
					startunbounded: true,
					stop: 100,
					stopinclusive: false,
					"label": "<100",
				}, {
					start: 100,
					startinclusive: true,
					stopunbounded: true,
					label: "≥100",
				}]
			} },
			//hidden:['exposed, dose unknown','unknown exposure']
		})],
		['Cardio/hrtavg/race', __cuminc({
			t0:{id:'racegrp'},
			t1:{id: 'Cardiovascular System'},
			t2:{id:'hrtavg', q: {
				type: "custom-bin",
				mode: "discrete",
				lst: [{
					startunbounded: true,
					stop: 100,
					stopinclusive: false,
					"label": "<100",
				}, {
					start: 100,
					startinclusive: true,
					stopunbounded: true,
					label: "≥100",
				}]
			} },
			//hidden:['exposed, dose unknown','unknown exposure']
		})],

		'matrix',

		['SJLife','?noheader=1&mass='+JSON.stringify({
			dslabel:'SJLife',
			nav: { activeTab: 1 },
			genome:'hg38',
			plots: [
				{
					chartType:'matrix',
					termgroups: [{
						name: 'Demographics',
						lst: [{
							id: 'sex'
						},{
							id: 'agedx'
						},{
							id: 'genetic_race'
						}]
					},{
						name: 'Treatment',
						lst: [{
							id: 'hrtavg'
						}]
					},{
						name: 'Chronic Conditions',
						lst: [{
							id: 'Arrhythmias'
						},{
							id: 'Renal System'
						},{
							id: 'Musculoskeletal System'
						}]
					}]
				}
			],
			termfilter: {
				filter: {"type":"tvslst","in":true,"join":"and","lst":[{"tag":"cohortFilter","type":"tvs","tvs":{"term":{"id":"subcohort","type":"categorical"},"values":[{"key":"SJLIFE","label":"SJLIFE"}]}},{"type":"tvslst","in":true,"join":"","lst":[{"type":"tvs","tvs":{"term":{"type":"categorical","values":{"Non-Hodgkin lymphoma":{"label":"Non-Hodgkin lymphoma"},"Hodgkin lymphoma":{"label":"Hodgkin lymphoma"},"Central nervous system (CNS)":{"label":"Central nervous system (CNS)"},"Acute myeloid leukemia":{"label":"Acute myeloid leukemia"},"Wilms tumor":{"label":"Wilms tumor"},"Acute lymphoblastic leukemia":{"label":"Acute lymphoblastic leukemia"},"Ewing sarcoma family of tumors":{"label":"Ewing sarcoma family of tumors"},"Neuroblastoma":{"label":"Neuroblastoma"},"Rhabdomyosarcoma":{"label":"Rhabdomyosarcoma"},"Osteosarcoma":{"label":"Osteosarcoma"},"Other leukemia":{"label":"Other leukemia"},"MDS/Acute myeloid leukemia":{"label":"MDS/Acute myeloid leukemia"},"Germ cell tumor":{"label":"Germ cell tumor"},"Soft tissue sarcoma":{"label":"Soft tissue sarcoma"},"Other malignancy":{"label":"Other malignancy"},"Chronic myeloid leukemia":{"label":"Chronic myeloid leukemia"},"Retinoblastoma":{"label":"Retinoblastoma"},"Melanoma":{"label":"Melanoma"},"Liver malignancies":{"label":"Liver malignancies"},"Nasopharyngeal carcinoma":{"label":"Nasopharyngeal carcinoma"},"Other carcinoma":{"label":"Other carcinoma"},"Non-malignancy":{"label":"Non-malignancy"},"Histiocytosis":{"label":"Histiocytosis"},"Colon carcinoma":{"label":"Colon carcinoma"}},"groupsetting":{"inuse":false},"name":"Diagnosis Group","id":"diaggrp","isleaf":true,"included_types":["categorical"],"child_types":[]},"values":[{"samplecount":95,"label":"Germ cell tumor","key":"Germ cell tumor","bar_width_frac":0.0729}]},"$id":6}],"tag":"filterUiRoot","$id":2}]}
			}
		})],

		['PNET','?noheader=1&mass='+JSON.stringify({
			dslabel:'PNET',
			genome:'hg19',
			plots: [
				{
					chartType:'matrix',
					termgroups: [{
						lst: [{
							term:{
								name:'TP53',
								type:'geneVariant'
							}
						}]
					}]
				}
			]
		})],

		['ASH','?noheader=1&mass='+JSON.stringify({
			dslabel:'ASH',
			genome:'hg38',
			plots: [
				{
					chartType:'matrix',
					termgroups: [{
						lst: [
						{ term:{
							name:'BCR',
							type:'geneVariant'
						} },
						{ term:{
							name:'KRAS',
							type:'geneVariant'
						} },
						{ term:{
							name:'NOTCH1',
							type:'geneVariant'
						} },
						]
					}]
				}
			]
		})],

		['GDC','?noheader=1&mass='+JSON.stringify({
			dslabel:'GDC',
			genome:'hg38',
			plots: [
				{
					chartType:'matrix',
					termgroups: [{
						lst: [
						{ term:{
								name:'MYC',
								type:'geneVariant'
						} },
						{ term:{
								name:'HOXA1',
								type:'geneVariant'
						} },
						// cannot use AKT1 for now due to incompatible refseq canonical isoform
						]
					}],
					settings: {
						matrix: {
							colw: 8
						}
					}
				}
			]
		})],
		'others',
		
		['scatter plot', '?noheader=1&mass={"dslabel":"PNET","genome":"hg19","plots":[{"chartType":"sampleScatter","name":"Methylome TSNE","colorTW":{"id":"TSNE Category"}}]}'],

		['survival','?noheader=1&mass='+JSON.stringify( {
			dslabel:'TermdbTest',
			nav: { activeTab: 1 },
			genome:'hg38',
			plots: [ {
				chartType:'survival',
				term2: { id: 'diaggrp' },
				term: { id: 'efs' }
			} ],
		})],


		['num filter','?noheader=1&mass='+JSON.stringify( {
			dslabel:'SJLife',
			genome:'hg38',
			nav: { activeTab:2, },
			termfilter: {
				filter:{
					type:'tvslst',
					join:'',
					in:true,
					lst:[
						{type:'tvs',tvs: {
							term: {id: "aaclassic_5", name: "Cumulative Alkylating Agent (Cyclophosphamide Equivalent Dose)", type:'float',
								"values":{
									"0": { "label":"Not exposed", "uncomputable":true },
									"-8888": { "label":"Exposed but dose unknown", "uncomputable":true },
									"-9999": { "label":"Unknown treatment record", "uncomputable":true }
								}
							},
							ranges: [
								{ value: '0', label: "Not exposed"},
								{startunbounded:true,stop:2000,stopinclusive:true}
							]
						}}
					]
				}
			},
		})],

		['!num filter','?noheader=1&mass='+JSON.stringify( {
			dslabel:'SJLife',
			genome:'hg38',
			nav: { activeTab:2, },
			termfilter: {
				filter:{
					type:'tvslst',
					join:'',
					in:true,
					lst:[
						{type:'tvs',tvs: {
							term: {id: "aaclassic_5", name: "Cumulative Alkylating Agent (Cyclophosphamide Equivalent Dose)", type:'float',
								"values":{
									"0": { "label":"Not exposed", "uncomputable":true },
									"-8888": { "label":"Exposed but dose unknown", "uncomputable":true },
									"-9999": { "label":"Unknown treatment record", "uncomputable":true }
								}
							},
							ranges: [
								{start:0, startinclusive:true,stopunbounded:true}
							],
							isnot:true
						}}
					]
				}
			},
		})],
	]
},
{name:section_regression,
	links:[

		'LINEAR',

		['EF', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [
				{
					chartType: 'regression',
					regressionType: 'linear',
					outcome: {
						id: 'LV_Ejection_Fraction_3D'
					},
				},
			]
		})],

		['EF ~ anthracycline', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [
				{
					chartType: 'regression',
					regressionType: 'linear',
					outcome: {
						id: 'LV_Ejection_Fraction_3D',
					},
					independent: [
						{
							// adding varclass to independent term will break at termsetting.js line 417. why??
							id: 'anthracyclines_jco_5',
							q: {
								mode:'discrete',
								type:'custom-bin',
								lst:[
									{startunbounded:true, stopinclusive:true,stop:0,label:'Not exposed'},
									{stopinclusive:true,start:0,stop:100,label:'>0 to 100'},
									{stopinclusive:true,start:100,stop:250,label:'>100 to 250'},
									{startinclusive:false,stopunbounded:true,start:250,label:'>250'}
								]
							},
							refGrp:'Not exposed'
						},
					]
				},
			]
		})],

		['EF ~ agedx diaggrp', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [
				{
					chartType: 'regression',
					regressionType: 'linear',
					outcome: {
						id: 'LV_Ejection_Fraction_3D',
					},
					independent: [
						{ id: "agedx"},
						{ id: "diaggrp"}
					]
				},
			]
		})],

		['EF ~ agedxSpline diaggrp', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [
				{
					chartType: 'regression',
					regressionType: 'linear',
					outcome: {
						id: 'LV_Ejection_Fraction_3D',
					},
					independent: [
						{ id: "agedx", q:{mode:'spline',knots:[{value:1},{value:4.5},{value:8.64},{value:17}]}},
						{ id: "diaggrp"}
					]
				},
			]
		})],

		['EF ~ agedx*sex', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [
				{
					chartType: 'regression',
					regressionType: 'linear',
					outcome: { id: 'LV_Ejection_Fraction_3D' },
					independent: [
						{ id: 'agedx', interactions:['sex']},
						{ id: 'sex', interactions:['agedx'], refGrp:'2'}
					]
				},
			]
		})],
		['EF ~ snplst sex', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'linear',
				outcome: { id: 'LV_Ejection_Fraction_3D' },
				independent: [ snplstTerm, { id: 'sex', refGrp:'2'} ]
			} ]
		})],

		['EF ~ sex snplocus', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'linear',
				outcome: { id: 'LV_Ejection_Fraction_3D' },
				independent: [
					{ id: 'sex', refGrp:'2'},
					{
						term:{ type:'snplocus', },
						q:snplocusQ
					}
				]
			} ]
		})],

		['EF ~ snplocus*sex*agedx', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'linear',
				outcome: { id: 'LV_Ejection_Fraction_3D' },
				independent: [
					{
						term:{ type:'snplocus', id:'temp112233' },
						interactions:['sex','agedx'],
						q:snplocusQ
					},
					{ id: 'sex', refGrp:'2', interactions:['temp112233']},
					{ id: 'agedx', interactions:['temp112233']}
				]
			} ]
		})],

		['EF ~ snplocus agedxSpline', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'linear',
				outcome: { id: 'LV_Ejection_Fraction_3D' },
				independent: [
					{
						term:{ type:'snplocus' },
						q:snplocusQ
					},
					{ id: "agedx", q:{mode:'spline',knots:[{value:1},{value:4.5},{value:8.64},{value:17}]}}
				]
			} ]
		})],

		['Yadav EF 2021','?noheader=1&mass='+JSON.stringify({
			genome:'hg38',
			dslabel:'SJLife',
			nav: { activeTab: 1 },
			plots:[{
				chartType: 'regression',
				regressionType: 'linear',
				outcome: { id: 'LV_Ejection_Fraction_3D' },
				independent: [
					{id:'sex'},
					{id:'agedx'},
					{id:'agelastvisit'},
					{id:'anthracyclines_cog_5',q:{
						mode:'discrete',type:'custom-bin',lst:[
							{startunbounded:true, stopinclusive:true,stop:0,label:'Not exposed'},
							{stopinclusive:true,start:0,stop:250,label:'>0 to 250'},
							{startinclusive:false,stopunbounded:true,start:250,label:'>250'}
						]
					}},
					{id:'hrtavg',q:{
						mode:'discrete',type:'custom-bin',lst:[
							{startunbounded:true, stopinclusive:true,stop:200,label:'Lower than 200'},
							{stopinclusive:true,start:200,stop:2000,label:'>200 to 2000'},
							{startinclusive:false,stopunbounded:true,start:2000,label:'>2000'}
						]
					}},
					{
						term:{ type:'snplocus', },
						q: (()=>{const a = JSON.parse(JSON.stringify(snplocusQ));a.chr='chr1';a.start=113558796;a.stop=113559396;return a})()
					}
				]
			}]
		})],

		'LOGISITC',

		['EF', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [{
				chartType: 'regression',
				regressionType: 'logistic',
				outcome: {
					id: 'LV_Ejection_Fraction_3D',
				}
			}]
		})],

		['EF ~ agedx diaggrp', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [
				{
					chartType: 'regression',
					regressionType: 'logistic',
					outcome: { id: 'LV_Ejection_Fraction_3D', },
					independent: [
						{ id: "agedx"},
						{ id: "diaggrp"}
					]
				},
			]
		})],

		['EF ~ agedxSpline diaggrp', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [
				{
					chartType: 'regression',
					regressionType: 'logistic',
					outcome: {
						id: 'LV_Ejection_Fraction_3D',
					},
					independent: [
						{ id: "agedx", q:{mode:'spline',knots:[{value:1},{value:4.5},{value:8.64},{value:17}]}},
						{ id: "diaggrp"}
					]
				},
			]
		})],

		['EF ~ diaggrp hrtavg', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [{
				chartType: 'regression',
				regressionType: 'logistic',
				outcome: {
					id: 'LV_Ejection_Fraction_3D',
					q: {
						"mode":"binary",
						"type":"custom-bin",
						"lst":[
							{"startunbounded":true,"stopinclusive":true,"stop":57.8,"label":"≤57.8"},
							{"stopunbounded":true,"startinclusive":false,"start":57.8,"label":">57.8"}
						]
					},
					"refGrp":"≤57.8"
				},
				independent: [
					{ id: "diaggrp"},
					{id:"hrtavg"},
					{
						"id":"genetic_race",
						"q":{
							"hiddenValues":{},
							"groupsetting":{"inuse":false}
						},
						"refGrp":"European Ancestry",
						"type":"categorical"
					}
				]
			}]
		})],

		['EF ~ snplst sex', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'logistic',
				outcome: { id: 'LV_Ejection_Fraction_3D' },
				independent: [ snplstTerm, { id: 'sex', refGrp:'2'} ]
			} ]
		})],

		['EF ~ sex snplocus', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'logistic',
				outcome: { id: 'LV_Ejection_Fraction_3D' },
				independent: [
					{ id: 'sex', refGrp:'2'},
					{
						term:{ type:'snplocus', },
						q:snplocusQ
					}
				]
			} ]
		})],

		['EF ~ snplocus*sex', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'logistic',
				outcome: { id: 'LV_Ejection_Fraction_3D' },
				independent: [
					{
						term:{ type:'snplocus', id:'temp112233' },
						interactions:['sex'],
						q:snplocusQ
					},
					{ id: 'sex', refGrp:'2', interactions:['temp112233']}
				]
			} ]
		})],

		['Arrhythmias', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [{
				chartType: 'regression',
				regressionType: 'logistic',
				outcome: { id: 'Arrhythmias' }
			}]
		})],

		['Arrhythmias ~ chestrt', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [{
				chartType: 'regression',
				regressionType: 'logistic',
				outcome: {
					id: 'Arrhythmias',
					q:{ breaks:[2],groupNames:['No condition / not tested','Has condition']},
					refGrp:'No condition / not tested'
				},
				independent: [
					{
						id:"chestrt_yn",
						q: {
							mode:'discrete',
							type:'custom-bin',
							lst:[
								{startunbounded:true, stopinclusive:true,stop:0,label:'Not exposed'},
								// FIXME no need to define startinclusive
								{startinclusive:false,stopinclusive:true,start:0,stop:100,label:'>0 to 100'},
								{startinclusive:false,stopunbounded:true,start:100,label:'>100'}
							]
						},
						refGrp:'Not exposed'
					},
				]
			}]
		})],

		['BreastSN ~ PGS + age + chestRT', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [{
				chartType: 'regression',
				regressionType: 'logistic',
				outcome: {
					id:'Breast',
					q:{ breaks:[2],groupNames:['No condition / not tested','Has condition']},
					refGrp:'No condition / not tested'
				},
				independent:[
					{
						id:'prs_PGS000332',
						q:{
							"mode": "discrete",
							"scale": 1,
							"type": "regular-bin",
							"startinclusive": true,
							"bin_size": 0.4,
							"first_bin": {
								"stop": -1,
								"bin": "first",
								"startunbounded": true
							},
							"rounding": ".1f",
							"termtype": "float",
						},
						refGrp:'-0.2 to <0.2'
					},
					{id:'agelastvisit'},
					{id:'chestrt_yn',
						q: {
							"type": "custom-bin",
							"lst": [
								{
									"startunbounded": true,
									"stop": 0,
									"startinclusive": false,
									"stopinclusive": true,
									"label": "Not exposed"
								},
								{
									"start": 0,
									"startinclusive": false,
									"stopinclusive": true,
									"stop": 1000,
									"label": ">0 to 1000"
								},
								{
									"start": 1000,
									"startinclusive": false,
									"stopinclusive": true,
									"stopunbounded": true,
									"label": ">1000"
								}
							],
							"mode": "discrete"
						},
						refGrp:'Not exposed'
					}
				]
			}]
		})],

		['Arrhythmias ~ snplst sex', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [{
				chartType: 'regression',
				regressionType: 'logistic',
				outcome: { id: 'Arrhythmias' },
				independent: [ snplstTerm, { id: 'sex', refGrp:'2'} ]
			} ]
		})],

		['Arrhythmias ~ sex snplocus', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'logistic',
				outcome: {
					id: 'Arrhythmias',
					q:{ breaks:[2],groupNames:['No condition / not tested','Has condition']},
					refGrp:'No condition / not tested'
				},
				independent: [
					{ id: 'sex', refGrp:'2'},
					{
						term:{ type:'snplocus', },
						q: snplocusQ
					}
				]
			} ]
		})],

		['Arrhythmias ~ snplocus*sex', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'logistic',
				outcome: {
					id: 'Arrhythmias',
					q:{ breaks:[2],groupNames:['No condition / not tested','Has condition']},
					refGrp:'No condition / not tested'
				},
				independent: [
					{
						term:{ type:'snplocus', id:'temp112233' },
						interactions:['sex'],
						q:snplocusQ
					},
					{ id: 'sex', refGrp:'2', interactions:['temp112233']}
				]
			} ]
		})],

		// paper: https://pubmed.ncbi.nlm.nih.gov/31462438/
		['Stroke ~ snplocus','?mass-session-id=stroke&noheader=1'],

		'COX',

		['Neuro ~ sex', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'cox',
				outcome: { id:'Neurology System' },
				independent:[
					{ id: 'sex', refGrp:'2'}
				]
			}]
		})],

		['Cardio ~ chestrt', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'cox',
				outcome: { id:'Cardiovascular System', q:{breaks:[3]} },
				independent:[
					{
						id:"chestrt_yn",
						q: {
							mode:'discrete',
							type:'custom-bin',
							lst:[
								{startunbounded:true, stopinclusive:true,stop:0,label:'Not exposed'},
								{startinclusive:false,stopinclusive:true,start:0,stop:100,label:'>0 to 100'},
								{startinclusive:false,stopunbounded:true,start:100,label:'>100'}
							]
						},
						refGrp:'Not exposed'
					},
				]
			}]
		})],

		['Cardio ~ sex*agedx', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'cox',
				outcome: { id:'Cardiovascular System' },
				independent:[
					{ id:"sex", interactions:['agedx']},
					{ id:"agedx", interactions:['sex']},
				]
			}]
		})],

		['Neuro ~ snplst sex', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [{
				chartType: 'regression',
				regressionType: 'cox',
				outcome: { id: 'Neurology System' },
				independent: [ snplstTerm, { id: 'sex', refGrp:'2'} ]
			} ]
		})],

		['Neuro ~ sex snplocus', '?noheader=1&mass='+JSON.stringify({
			genome: 'hg38',
			dslabel: 'SJLife',
			nav: { activeTab: 1 },
			plots: [ {
				chartType: 'regression',
				regressionType: 'cox',
				outcome: {
					id: 'Neurology System', q:{ breaks:[2]},
				},
				independent: [
					{ id: 'sex', refGrp:'2'},
					{
						term:{ type:'snplocus', },
						q: snplocusQ
					}
				]
			} ]
		})],
		// https://pubmed.ncbi.nlm.nih.gov/33288658/
		['Yadav Cardiomyopathy 2021','?noheader=1&mass='+JSON.stringify({
			genome:'hg38',
			dslabel:'SJLife',
			nav: { activeTab: 1 },
			plots:[{
				chartType: 'regression',
				regressionType: 'cox',
				outcome: {
					id: 'Cardiomyopathy',
					q:{timeScale:'age'}
				},
				independent: [
					{id:'sex'},
					{id:'agedx'},
					{id:'agelastvisit'},
					{id:'anthracyclines_cog_5',q:{
						mode:'discrete',type:'custom-bin',lst:[
							{startunbounded:true, stopinclusive:true,stop:0,label:'Not exposed'},
							{stopinclusive:true,start:0,stop:250,label:'>0 to 250'},
							{startinclusive:false,stopunbounded:true,start:250,label:'>250'}
						]
					}},
					{id:'hrtavg',q:{
						mode:'discrete',type:'custom-bin',lst:[
							{startunbounded:true, stopinclusive:true,stop:200,label:'Lower than 200'},
							{stopinclusive:true,start:200,stop:2000,label:'>200 to 2000'},
							{startinclusive:false,stopunbounded:true,start:2000,label:'>2000'}
						]
					}},
					// chr1:113558796-113559396 for rs6689879
					// chr15:87090755-87090824 for rs9788776
					{
						term:{ type:'snplocus', },
						q: (()=>{const a = JSON.parse(JSON.stringify(snplocusQ));a.chr='chr1';a.start=113558796;a.stop=113559396;return a})()
					}
				]
			}]
		})],
	]
},
])

function makeURL(rows) {
	const div = document.getElementById('contentDiv')
	div.innerHTML = '' // clear text

/*
	const headers = ['', 'App Card', 'Links']
	for (const header of headers) {
		const d = document.createElement('div')
		d.innerHTML = header
		d.style.fontWeight = '550'
		div.appendChild(d)
	}
	*/

	for(const row of rows) {

		{
			///////////////////////////////////////
			// column 1: header
			///////////////////////////////////////

			const d = document.createElement('div')
			d.innerHTML = row.name || '.name missing'
			div.appendChild(d)
		}

		{
			///////////////////////////////////////
			// column 2: card link
			///////////////////////////////////////


			const d = document.createElement('div')
			div.appendChild(d)
			if(row.card) {
				const a = document.createElement('a')
				a.innerHTML = 'card'
				a.setAttribute('href', `${window.location.origin}/?appcard=${row.card}`)
				a.setAttribute('target', '_blank')
				d.appendChild(a)
			}

		}

		{
			///////////////////////////////////////
			// column 3: links
			///////////////////////////////////////
			const d = document.createElement('div')
			div.appendChild(d)

			if(!Array.isArray(row.links)) {
				d.innerHTML = '.links[] missing'
				continue
			}

			for(const [i,ele] of row.links.entries()) {
				if(typeof ele == 'string') {
					if(i!=0) {
						// only create <br> if the string is not the first in array
						d.append(document.createElement('br'))
					}
					const s = document.createElement('span')
					s.style.marginRight = '8px'
					s.style.opacity = '.5'
					s.innerHTML = ele
					d.appendChild(s)
					continue
				}
				if(Array.isArray(ele)) {
					// ele is [ name, link ]
					const a = document.createElement('a')
					a.innerHTML = ele[0]
					a.setAttribute('href',window.location.origin+'/'+ele[1])
					a.setAttribute('target','_blank')
					a.style.paddingRight = '8px'
					d.appendChild(a)
				} else {
					window.alert('invalid element: '+ele)
				}
			}
		}
	}
}

</script>

<div style="margin-top:30px;color:#858585">

<p>Above links use files from "tp/proteinpaint_demo/" folder on your system, and share files with App Drawer and <code>server/cards/index.json</code>.

<br>

These test files should be publicly sharable (with patient identifier removed etc), and should be kept in sync between your computer, hpc, and prp1.

<br>

A tarball of the entire proteinpaint_demo/ folder is available at https://pecan.stjude.cloud/static/pp-support/pp.demo.tgz</p>

<p>
README listing <a href=/readme.html>/readme.html</a>.
Browser tests <a href=/testrun.html>/testrun.html</a>.
Server health <a href=/healthcheck>/healthcheck</a>.
</p>

<p>TODO how to open JS console automatically in the new tab by clicking a link? That helps with debugging.</p>
</div>

</body>
</html>
